helper/helper/src/main/java/com/linkedin/avroutil1/compatibility/AvroRecordUtil.java

```diff
@@ -811,4 +820,147 @@ private static Enum<?> getSpecificEnumSymbol(Class<? extends Enum<?>> enumClass,
+            path.appendPath(new UnionTypePredicate(valueSchema.getName()));
+            try {
+              validateNestedValue(fieldValue, path, visited, referenceDefinitions, comparisonConfiguration);
+            } finally {
+              path.pop();
+            }
+            break;
+          case ARRAY:
+            Collection<?> arr = (Collection<?>) fieldValue;
+            Iterator<?> iter = arr.iterator();
+            int i=0;
+            while (iter.hasNext()) {
+              Object value = iter.next();
+              path.appendPath(new ArrayPositionPredicate(i));
+              try {
+                validateNestedValue(value, path, visited, referenceDefinitions, comparisonConfiguration);
+              } finally {
+                path.pop();
+                i++;
+              }
+            }
+            break;
+          case MAP:
+            @SuppressWarnings("unchecked")
+            Map<CharSequence, ?> map = (Map<CharSequence, ?>) fieldValue;
+            for (Map.Entry<CharSequence, ?> entry : map.entrySet()) {
+              CharSequence key = entry.getKey();
+              Object value = entry.getValue();
+              path.appendPath(new MapKeyPredicate(key));
+              try {
+                validateNestedValue(value, path, visited, referenceDefinitions, comparisonConfiguration);
+              } finally {
+                path.pop();
+              }
+            }
+            break;
+          default:
+            validateNestedValue(fieldValue, path, visited, referenceDefinitions, comparisonConfiguration);
+        }
+      } finally {
+        path.pop();
+      }
+    }
+  }
+
+  private static void validateNestedValue(
+          Object value,
+          AvroPath path,
+          IdentityHashMap<Schema, Boolean> visited,
+          Map<String, Schema> referenceDefinitions,
+          SchemaComparisonConfiguration comparisonConfiguration
+  ) throws InconsistentSchemaException {
+    if (value == null) {
+      return;
+    }
+    if (value instanceof IndexedRecord) {
+      //generic or specific record (dont care which)
+      IndexedRecord record = (IndexedRecord) value;
+      validateNestedRecordSchema(record, path, visited, referenceDefinitions, comparisonConfiguration);
+    } else {
+      Schema declaredSchema = AvroSchemaUtil.getDeclaredSchema(value);
+      if (declaredSchema != null) {
+        validateNestedSchema(declaredSchema, path, visited, referenceDefinitions, comparisonConfiguration);
+      }
+    }
+  }
+
+  private static void validateNestedSchema(
+          Schema schema,
+          AvroPath path,
+          IdentityHashMap<Schema, Boolean> visited,
+          Map<String, Schema> referenceDefinitions,
+          SchemaComparisonConfiguration comparisonConfiguration
+  ) throws InconsistentSchemaException {
+    if (!visited.containsKey(schema)) {
+      visited.put(schema, Boolean.TRUE); //never again
```

Comment:
Can the key here be a fullname string instead of a schema? or do we have to use the schema because of different schemas with the same fullname?

Reply:
the issue this is meant to expose is different definitions of the "same" schema. caching on schema name would prevent this from working. this is an identity cache (on instances of org.apache.avro.Schema) for that reason